### PR TITLE
[Semio] Scratch Support

### PR DESCRIPTION
**Pull request by bmcdorman**
# Overview
This PR adds Scratch support to the KIPR Simulator. It does this by embedding a highly customized [`scratch-blocks` ](https://github.com/LLK/scratch-blocks) instance into the simulator. This customized version is stored at [`kipr-scratch`](https://github.com/kipr/kipr-scratch). A custom VM is leveraged to run programs: [`itch`](https://github.com/kipr/itch). A C program is compiled as a part of static simulator dependencies and "hijacked" by itch to execute KIPR functions natively inside scratch.

## List of Changes

1. "Scratch" language option added to Editor
2. Scratch programs saved like existing programs.
3. Highly customized scratch editor integrated. See `ScratchEditor.tsx` and `kipr-scratch`.
4. Custom runtime developed via itch and `scratch-rt.js` (see dependencies)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -25,8 +25,6 @@ sudo apt-get install nodejs cmake build-essential swig python3.8
 sudo npm install --global npm
```
Comment (navzam) [inline README.md:25]: "wait_for_milliseconds" block isn't working.
Console output:
> No module registered for block type wait

Comment (navzam) [reply README.md:25]: Same for "wait_for_touch" block

Comment (navzam) [inline README.md:25]: "wait x seconds" block isn't working.
Console output:
> No function registered for block type control_wait

**Comment by tcorbly** [comment]
Program will not run anything else after repeat until block.
Screenshot from 2023-02-21 16-01-30

**Comment by tcorbly** [comment]
We probably want a more distinguished line between block options and the editor space.

**Comment by tcorbly** [comment]
Issue related to either the `gmpc` or `repeat until`. The program throw errors when using a variable as part of the condition and also does not appear to either receive an output from `gmpc` or is skipping it for some reason.
Screenshot from 2023-02-21 23-15-47

File: README.md (hunk 1 above)
Comment (navzam) [reply README.md:25]: Reopening b/c block still isn't working. Same output.
"wait_for_milliseconds" and "wait_for_touch" are working though

[the remark this answers is not in the document]
Comment (navzam) [reply README.md:25]: Same with "if then else" block

Comment (navzam) [inline README.md:25]: I think the "repeat until" block is reversed. This runs forever:
image

while reversing the condition terminates immediately

Comment (navzam) [inline README.md:25]: "=" block gives error:
> No function registered for block type operator_equals

Comment (navzam) [inline README.md:25]: "pick random x to y" block behaves kinda strange if x >= y. I'm not sure what the behavior should be, but I think it should be consistent (i.e. either include the endpoints or don't)

| x | y | output |
| - | - | - |
| 4 | 4 | [4] |
| 4 | 3 | [4] |
| 4 | 2 | [3] |
| 4 | 1 | [2, 3] |
| 4 | 0 | [1, 3] |

Comment (navzam) [inline README.md:25]: "operator of x" block gives error:
> OPERATOR not found

Happens for all operators, not just "abs"
image

Comment (bmcdorman) [reply README.md:25]: I'll make it an error

Comment (navzam) [inline README.md:25]: "show variable x"/"hide variable x" don't work in a way that's useful to the user

Comment (navzam) [inline README.md:25]: "repeat until" block condition is still backwards, but only when the block is empty

Comment (navzam) [inline README.md:25]: "change x by y" block fails with output:
> VARIABLE must be number

image

```xml
<xml xmlns="http://www.w3.org/1999/xhtml">
  <variables>
    <variable type="" id=";/fOvI=r,rhuvz3kg:!y" islocal="false" iscloud="false">i</variable>
    <variable type="" id="mXKmvt*Z7vJ|3:~+HU:#" islocal="false" iscloud="false">j</variable>
    <variable type="" id="2ogUd*aq6_gfGWacz#/2" islocal="false" iscloud="false">testvar</variable>
    <variable type="" id="q=qhREa7eU=jdV,%uUs-" islocal="false" iscloud="false">k</variable>
  </variables>
  <block type="control_run" id="0fNAs-vY8^BL;[T$]eav" x="159" y="8">
    <next>
      <block type="data_setvariableto" id="(RpP{VpAeA!KNbm}Tscv">
        <field name="VARIABLE" id="2ogUd*aq6_gfGWacz#/2" variabletype="">testvar</field>
        <value name="VALUE">
          <shadow type="text" id="2#bAQ,?kgq|Jf~}kEU$P">
            <field name="TEXT">4</field>
          </shadow>
        </value>
        <next>
          <block type="data_changevariableby" id="Ed+~U!%,|,PK^ef%1JHG">
            <field name="VARIABLE" id="2ogUd*aq6_gfGWacz#/2" variabletype="">testvar</field>
            <value name="VALUE">
              <shadow type="math_number" id=";T;+tWn,?Zbt}mK,j/0C">
                <field name="NUM">1</field>
              </shadow>
            </value>
          </block>
        </next>
      </block>
    </next>
  </block>
</xml>
```

Comment (navzam) [inline README.md:25]: Using a variable in the "=" block fails with output:
> Invalid operand type

image

```xml
<xml xmlns="http://www.w3.org/1999/xhtml">
  <variables>
    <variable type="" id=";/fOvI=r,rhuvz3kg:!y" islocal="false" iscloud="false">i</variable>
    <variable type="" id="mXKmvt*Z7vJ|3:~+HU:#" islocal="false" iscloud="false">j</variable>
    <variable type="" id="2ogUd*aq6_gfGWacz#/2" islocal="false" iscloud="false">testvar</variable>
    <variable type="" id="q=qhREa7eU=jdV,%uUs-" islocal="false" iscloud="false">k</variable>
    <variable type="" id="m+BqD().ru``m17|I[)," islocal="false" iscloud="false">m</variable>
  </variables>
  <block type="control_run" id="0fNAs-vY8^BL;[T$]eav" x="159" y="8">
    <next>
      <block type="data_setvariableto" id="(RpP{VpAeA!KNbm}Tscv">
        <field name="VARIABLE" id="2ogUd*aq6_gfGWacz#/2" variabletype="">testvar</field>
        <value name="VALUE">
          <shadow type="text" id="2#bAQ,?kgq|Jf~}kEU$P">
            <field name="TEXT">4</field>
          </shadow>
        </value>
        <next>
          <block type="control_if" id="eU;`eR~$4=MUK7b9sG0E">
            <value name="CONDITION">
              <block type="operator_equals" id="=%$4Pc2{Ju]u:ag%L37H">
                <value name="OPERAND1">
                  <shadow type="text" id="zNi~dy7NAcJNT~G-;gx8">
                    <field name="TEXT">1</field>
                  </shadow>
                  <block type="data_variable" id="BCKI$6kiG|4c/|}Ob2ul">
                    <field name="VARIABLE" id="2ogUd*aq6_gfGWacz#/2" variabletype="">testvar</field>
                  </block>
                </value>
                <value name="OPERAND2">
                  <shadow type="text" id="OUt;O2{3..-Mk6*]LPCe">
                    <field name="TEXT">1</field>
                  </shadow>
                </value>
              </block>
            </value>
          </block>
        </next>
      </block>
    </next>
  </block>
</xml>
```

Comment (navzam) [inline README.md:25]: "letter x of y" block fails with output:
> Invalid operand type

image

```xml
<xml xmlns="http://www.w3.org/1999/xhtml">
  <variables>
    <variable type="" id=";/fOvI=r,rhuvz3kg:!y" islocal="false" iscloud="false">i</variable>
    <variable type="" id="mXKmvt*Z7vJ|3:~+HU:#" islocal="false" iscloud="false">j</variable>
    <variable type="" id="2ogUd*aq6_gfGWacz#/2" islocal="false" iscloud="false">testvar</variable>
    <variable type="" id="q=qhREa7eU=jdV,%uUs-" islocal="false" iscloud="false">k</variable>
    <variable type="" id="m+BqD().ru``m17|I[)," islocal="false" iscloud="false">m</variable>
    <variable type="" id="#S;ca-yM_;tly+p4=K4o" islocal="false" iscloud="false">n</variable>
    <variable type="" id="zxEcNq=cqSYy+euCDP32" islocal="false" iscloud="false">o</variable>
  </variables>
  <block type="control_run" id="0fNAs-vY8^BL;[T$]eav" x="159" y="8">
    <next>
      <block type="data_setvariableto" id="4cVTC%HH)byt|f:g#n~c">
        <field name="VARIABLE" id="2ogUd*aq6_gfGWacz#/2" variabletype="">testvar</field>
        <value name="VALUE">
          <shadow type="text" id="%L,.yYJ%#uBN@-Mpj(98">
            <field name="TEXT">0</field>
          </shadow>
          <block type="operator_letter_of" id="AV*Z(ob+RcU+ZZ*LUl~I">
            <value name="LETTER">
              <shadow type="math_whole_number" id="%!sO9QkRua:yT1S-I.^M">
                <field name="NUM">1</field>
              </shadow>
            </value>
            <value name="STRING">
              <shadow type="text" id="3lPB[X8`PoL2s2XX4r0*">
                <field name="TEXT">world</field>
              </shadow>
            </value>
          </block>
        </value>
      </block>
    </next>
  </block>
</xml>
```

Comment (navzam) [inline README.md:25]: Clicking "Reset" in the Scratch editor doesn't delete variables from the UI variable list. The variables do seem to be deleted based on the XML, and the UI eventually updates when other variables are changed

File: express.js [hunk 2]
```diff
@@ -88,7 +89,7 @@ if (config.server.dependencies.libkipr_c && config.server.dependencies.emsdk_env
       env['EMSDK'] = config.server.dependencies.emsdk_env.EMSDK;
       env['EM_CONFIG'] = config.server.dependencies.emsdk_env.EM_CONFIG;
   
-      exec(`emcc -s WASM=0 -s INVOKE_RUN=0 -s ASYNCIFY -s EXIT_RUNTIME=1 -s "EXPORTED_FUNCTIONS=['_main', '_simMainWrapper']" -I${config.server.dependencies.libkipr_c}/include -L${config.server.dependencies.libkipr_c}/lib -lkipr -o ${path}.js ${path}`, {
+      exec(`emcc -s WASM=0 -s INVOKE_RUN=0 -s ASYNCIFY -s EXIT_RUNTIME=1 -s "EXPORTED_FUNCTIONS=['_main', '_simMainWrapper']" -I${config.server.dependencies.libkipr_c}/include -Wl,--whole-archive -L${config.server.dependencies.libkipr_c}/lib -lkipr -o ${path}.js ${path}`, {
```
Comment (navzam) [inline express.js:92]: Is `--whole-archive` needed? I think it bloats C compilation.
A program with just a `motor` command is 5 MB with it and 4.4 MB without it.
A "hello world" program with no libwallaby references is 5 MB with it and 470 kB without it.

Comment (bmcdorman) [reply express.js:92]: This was accidentally committed. Good catch.

File: README.md (hunk 1 above)
Comment (navzam) [inline README.md:25]: Clicking "Run" with an empty program gives an unhelpful error message in the sim console:
> Cannot read properties of undefined (reading 'documentElement')

I think it only happens the very first time you use Scratch. After that (when the Scratch program isn't a completely empty string), it shows the more useful "block not found" error.

Dev console shows:
```
dom-parser.js:178 
        
       [xmldom error]	invalid doc source 
@#[line:0,col:undefined]
error @ dom-parser.js:178
DOMParser.parseFromString @ dom-parser.js:27
mod.onRuntimeInitialized @ worker.ts:210
doRun @ VM5:156756
run @ VM5:156773
runCaller @ VM5:156681
removeRunDependency @ VM5:136060
receiveInstance @ VM5:136223
receiveInstantiationResult @ VM5:136241
Promise.then (async)
instantiateArrayBuffer @ VM5:136249
instantiateAsync @ VM5:136293
createWasm @ VM5:136311
eval @ VM5:141483
__WEBPACK_DEFAULT_EXPORT__ @ require.ts:9
(anonymous) @ worker.ts:178
fulfilled @ worker.ts:10
Promise.then (async)
step @ worker.ts:12
fulfilled @ worker.ts:10
Promise.then (async)
step @ worker.ts:12
(anonymous) @ worker.ts:13
__awaiter @ worker.ts:9
startScratch @ worker.ts:164
(anonymous) @ worker.ts:283
(anonymous) @ worker.ts:13
__awaiter @ worker.ts:9
start @ worker.ts:261
ctx.onmessage @ worker.ts:292
worker.ts:242 
        
       TypeError: Cannot read properties of undefined (reading 'documentElement')
    at Ast.parse (Ast.ts:121:1)
    at InstanceIncompleteOptions.complete (Instance.ts:6:1)
    at new Instance (Instance.ts:129:1)
    at mod.onRuntimeInitialized (worker.ts:209:1)
    at doRun (eval at __WEBPACK_DEFAULT_EXPORT__ (require.ts:9:1), <anonymous>:156756:71)
    at run (eval at __WEBPACK_DEFAULT_EXPORT__ (require.ts:9:1), <anonymous>:156773:5)
    at runCaller (eval at __WEBPACK_DEFAULT_EXPORT__ (require.ts:9:1), <anonymous>:156681:19)
    at removeRunDependency (eval at __WEBPACK_DEFAULT_EXPORT__ (require.ts:9:1), <anonymous>:136060:7)
    at receiveInstance (eval at __WEBPACK_DEFAULT_EXPORT__ (require.ts:9:1), <anonymous>:136223:5)
    at receiveInstantiationResult (eval at __WEBPACK_DEFAULT_EXPORT__ (require.ts:9:1), <anonymous>:136241:5)
```

Comment (navzam) [inline README.md:25]: Show/hide variable only works with a sleep. The program below will not show the variable. And even if you add a sleep after the "show variable" block, the variable is not shown until the end of the sleep period.

image

```
<xml xmlns="http://www.w3.org/1999/xhtml">
  <variables>
    <variable type="" id="!BBtuLB9KEV*u}`|z*2@" islocal="false" iscloud="false">test</variable>
    <variable type="" id="6fz|yAaemr1QU;[@_q@s" islocal="false" iscloud="false">i</variable>
    <variable type="" id="!bq%PTcdJnN1HU9`d$4Q" islocal="false" iscloud="false">j</variable>
    <variable type="" id="0-XOGI~a{yiQ_SB.$7fp" islocal="false" iscloud="false">k</variable>
    <variable type="" id="Xm?:86*Zayd!O*R5xR#:" islocal="false" iscloud="false">m</variable>
    <variable type="" id="=g$Rw,tGqz7QYDTx]UVJ" islocal="false" iscloud="false">n</variable>
  </variables>
  <block type="control_run" id="JorN:3{p^ujG@Qr`(,y#" x="137" y="153">
    <next>
      <block type="data_setvariableto" id="}IUiky{mG@/EbyhQ]gpD">
        <field name="VARIABLE" id="!BBtuLB9KEV*u}`|z*2@" variabletype="">test</field>
        <value name="VALUE">
          <shadow type="text" id="5qjD9hI6?wN$FT}]PUl]">
            <field name="TEXT">2</field>
          </shadow>
        </value>
        <next>
          <block type="data_showvariable" id="hbW#3fpt%yqf2AXyoT-l">
            <field name="VARIABLE" id="!BBtuLB9KEV*u}`|z*2@" variabletype="">test</field>
          </block>
        </next>
      </block>
    </next>
  </block>
</xml>
```

Comment (navzam) [reply README.md:25]: Re-opening. Still failing with the same error. I noticed that it fails with the program above, where the variable is set to "4", but not when the variable is set to a function returning a number, like `gmpc()`

Comment (navzam) [inline README.md:25]: This program outputs the variable twice. Not sure if it's intended behavior.

image

And this program outputs it 4 times:

image

Comment (chrismbirmingham) [reply README.md:25]: Show variable outputs the variable every second, to stop it from showing you can use hide variable. Seems to be the intended flow, though we could probably make changes if we choose to.

**Pull request closed without merging**